Adds security_testing Proposal (#235)
ros2/design: articles/ros2_sectest.md

---
layout: default
title: ROS 2 Security Testing
permalink: articles/ros2_sectest.html
abstract:
  This document describes a proposed package for testing
  security concerns specific to ROS 2.
author: >
  [Thomas Moulard](https://github.com/thomas-moulard),
  [Ryan Newell](https://github.com/ryanewel)

published: true
categories: Security
---

{:toc}

{::options parse_block_html="true" /}
<style>
table {
    table-layout: fixed;
    width: 100%;
}

th {
    width: 11em;
}
</style>

# {{ page.title }}

## Intent Of This Document
The intent of this document is to:

> 1. Describe the design of a new ROS 2 packages called `ROS2-SecTest`. This package will gather code which can be used to "attack" ROS 2 applications.

Reviewer: Since `ROS2-SecTest` isn't a valid package name please use something like `security_testing` instead. Also "ROS2" in the package name doesn't make much sense since every package is a ROS (2) package.

Author: Good point, changed the package name

> 1. Describe the design of a new ROS 2 packages called `ROS2-SecTest`. This package will gather code which can be used to "attack" ROS 2 applications.

Reviewer: Please follow the [style guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#markdown-restructured-text-docblocks): each sentence should start on a new line.

Author: sorry about that, updated

3. **Mitigation PoC** - the mitigation, from a practical standpoint, means an improvement (PR) contributed back to any ROS 2 package.
4. **Mitigation Validation** - Once the PR in 3. is merged, the step 2. should be repeated. At this point, the attack code should **not** be able to disrupt or impact the robot in any way. if the mitigation is partial (requires particular setup or configuration or does not cover all cases), this should be explicitly called out in the vulnerability report.

## Vulnerability Evaluation Process Scope

### ROS 2 specific VS Generic Vulnerabilities

Unlike the ALIAS Robotic Vulnerability Disclosure Program [RVDP][RVDP], this security assessment program focuses on vulnerabilities *directly related to ROS 2*.

For instance, the following issues are in scope:

* Leaking ROS 2 topic data,

> * Triggering ROS 2 services or actions without being unauthorized,

Reviewer 2: * Triggering ROS 2 services or actions without being authorized,

> * Triggering ROS 2 services or actions without being unauthorized,

Reviewer 2: ```suggestion
* Triggering ROS 2 services or actions without being authorized,
```

Reviewer 3: +1

* Abusing node API to disrupt the robot behavior.

However, vulnerabilities unrelated to ROS 2 are not in scope:

* Checking that the robot SSH server cannot be accessed using “easy to guess” credentials such as user = root, password = admin for instance.
* Validating that the robot NTP server or kernel is up-to-date and is not vulnerable to attacks.

The rationale is that any arbitrary software package may be bundled with a particular robot. Trying to evaluate all possible vulnerabilities can be done using a generic vulnerability scanner or framework. We should not try to re-invent generic tools already developed by the security community.

> On the opposite, we need to address ROS 2 specific concerns as limited efforts have been done to document and mitigate those issues.

Reviewer 2: Where do we draw the line e.g. what about a package depends on outdated library with known vulnerabilities? Or specific configuration issues?

Reviewer 4: Generally, both of our examples would be out of the scope of this package, unless there is something ROS specific to it.

E.g. an openssh exploit is not in scope. But if the way `roslaunch` invokes `ssh` makes the exploit more dangerous, this is something that would be in scope.

There may be edge cases. Those should be discussed on a case-by-case basis.

### Local VS Remote vulnerabilities

Both local and remote vulnerabilities *are in scope*.

A *remote* vulnerability is defined as a vulnerability which can be exploited from any remote computer which can communicate with the ROS 2 application running on the robot (usually, any host which can communicate with the robot using DDS).

A *local* vulnerability is defined as a vulnerability which can only be exploited from a robot computer.

While remote vulnerabilities are more critical than local vulnerabilities, we make the assumption that, in some cases, attacker will have access to a shell on the robot and that ROS 2 need to cope with those situations in the best way possible.

### Relationship to SROS

Attacks mitigated by SROS *are in-scope*.

SROS helps to prevent some attacks (unauthorized actor should not be able to read or write topic data. However, SROS may be disabled either accidentally, by an attacker or due to technical limitations (SROS overhead is unacceptable, etc.).

> As a consequence, we make the assumption that it is helpful to implement attacks for vulnerabilities which could be mitigated by SROS.

Reviewer: Without enabling security ROS (2) is by design open. So I am not sure how helpful it is to enumerate all the possible ways you can "attack" it. Obviously you can read all transmitted data, send arbitrary custom data, call every service / action with arbitrary input, etc.

Author: That's fair, I think this package could add value when developers are looking to move from an open ROS system to one with more closed permissions. The goal is that this package would give concrete and verifiable examples of the scenarios listed in the threat model and demonstrate how to mitigate them.

Reviewer 4: Some attacks are independent of whether or not SROS 2 is enabled. Either because they e.g. attack another aspect (resources) or because SROS 2 may not protect against it (DoS?).

Maybe it's better to rephrase it this way?

Reviewer 4: rephrased

## Intent Of This Document
The intent of this document is to:

1. Describe the design of a new ROS 2 packages called `security_testing`.
This package will gather code which can be used to "attack" ROS 2 applications.
1. Explain the motivation behind proposing this new package and how this work is a follow-up on the ROS 2 threat model.

We are contributing this document to the community to get buy-in from the community, raise visibility and hopefully create a single place for ROS 2 users to collaborate.

## Objective of Security Testing Package

We have previously published a document describing a [ROS 2 threat model][threat_model].
This threat model is composed of attacks which are, as of today, theoretical.
To continue raising the bar on ROS 2 security, we propose to:

1. Attempt to implement the attacks described in the document,
2. Propose and implement mitigations for the successful attacks.

The order in which threats will be considered will be determined by the threat vulnerability assessment score (DREAD) as well as how hard / how much time would be required to implement the attack.
We will target “low-hanging fruits” which can lead to critical vulnerabilities.

> 

Reviewer 3: ```suggestion
### Robot weaknesses and vulnerabilities
This document makes the following assumptions:
- A **bug** is an error, flaw, failure or fault in a computer program or system that causes it to produce an incorrect or unexpected result, or to behave in unintended ways.
- **Weaknesses** are errors that can lead to vulnerabilities. Bugs and weaknesses are treated indistinctly.
- A **vulnerability** is a weakness that can be exploited directly or indirectly by a hacker to produced unexpected behavior.

It's relevant to note that *all vulnerabilities are weaknesses but not the other way around*.
```

For commodity reasons, we suggest to use TurtleBot3 as a reference platform for this step.
Note: Not all attacks require validation on a robot.
A conscious effort should be made during this step to avoid over-relying on any particular robot specific characteristic.
If this is the case, those assumptions should be made explicit in the vulnerability report.
This attack validation needs to be easily reproducible by the community, but does not need to be automated (i.e. it is OK to ask the user to check manually whether the robot is moving differently due to the attack for instance).
3. **Mitigation PoC** - the mitigation, from a practical standpoint, means an improvement (PR) contributed back to any ROS 2 package.
4. **Mitigation Validation** - Once the PR in 3. is merged, the step 2. should be repeated.
At this point, the attack code should **not** be able to disrupt or impact the robot in any way.
If the mitigation is partial (requires particular setup or configuration or does not cover all cases), this should be explicitly called out in the vulnerability report.

## Vulnerability Evaluation Process Scope

### ROS 2 specific VS Generic Vulnerabilities

> Unlike the ALIAS Robotic Vulnerability Disclosure Program [RVDP][RVDP], this security assessment program focuses on vulnerabilities *directly related to ROS 2*.

Reviewer 3: ```suggestion
Unlike the Alias Robotics Robot Vulnerability Database [RVD][RVD], this security assessment program focuses on vulnerabilities *directly related to ROS 2*.
```

We would like to ensure this project becomes a part of the ROS 2 code base / ros2 organization on GitHub.
This is to ensure a high level of visibility to this project (and avoid duplication of efforts in the community).

## Selected Attacks to be implemented

To start, we chose to implement three “easy” attacks:

1. CPU abuse
2. Memory abuse
3. Disk abuse

Those three attacks need to run on the robot (local attacks).
They have been chosen because they are quick to implement and exhibit issues linked to the lack of sandboxing / isolation between ROS 2 nodes.

## Example of Attacks

The threat model is the source of truth, but we could also build the following exploits easily:

* (remote / SROS off) Record data from topics
* (remote / SROS off) Write data from topics
* (remote) Topic DDoS
* (local / SROS on) Time manipulation (node changes the system time, SROS fails because the deployed certificates are not valid for this point of time (e.g. if the SROS certificate indicates valid after May 2015, set the computer clock to 1970 to prevent any communication from happening).

[threat_model]: http://design.ros2.org/articles/ros2_threat_model.html

> [RVDP]: https://github.com/aliasrobotics/RVDP

Reviewer 3: ```suggestion
[RVD]: https://github.com/aliasrobotics/RVD
```
